examples/gaussian_process/svgp.py

```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+'''
+Sparse variational Gaussian process, loosely based on
+
+> Opper and Archambeau 2009, The variational Gaussian approximation revisited.
+
+On the Boston dataset you should get ~2.5 RMSE and ~2.3 negative log likelihood
+after a while with default args;
+On the protein dataset you should get ~4.5 RMSE and ~2.8 NLL after 100 epochs
+with `N_batch=1000 N_particles=10 dtype=float32`.
+'''
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import division
+
+from six.moves import range, zip
+
+import zhusuan as zs
+import tensorflow as tf
+import numpy as np
+from examples import conf
+from examples.utils import dataset
+import os
+import argparse
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-N_Z', default=100, type=int)
+parser.add_argument('-N_X', default=1, type=int)
+parser.add_argument('-N_particles', default=20, type=int)
+parser.add_argument('-N_particles_test', default=100, type=int)
+parser.add_argument('-N_batch', default=100, type=int)
+parser.add_argument('-N_epoch', default=5000, type=int)
+parser.add_argument('-dtype', default='float32', type=str,
+                    choices=['float32', 'float64'])
+parser.add_argument('-dataset', default='boston_housing', type=str,
+                    choices=['boston_housing', 'protein_data'])
+
+
+def rbf(x, y, inv_scale):
+    x = tf.expand_dims(x, 1)
+    y = tf.expand_dims(y, 0)
+    inv_scale = tf.reshape(inv_scale, [1, 1, -1])
+    return tf.exp(- tf.reduce_sum(tf.square(x - y) * inv_scale, axis=-1) / 2)
+
+
+@zs.reuse('model')
+def build_model(observed, Z_pos, X_ph, hps, n_particles, full_cov=False):
+    '''
+    Build the SVGP model.
+    Note that for inference, we only need the diagonal part of Cov[Y], as ELBO
+    equals sum over individual observations.
+    For visualization etc we may want a full covariance. Thus the argument
+    `full_cov`.
+    '''
+    with zs.BayesianNet(observed) as model:
+        k_inv_scale = tf.get_variable('k_inv_scale', [hps.N_X], hps.dtype,
+                                      tf.zeros_initializer())
+        k_inv_scale = tf.exp(k_inv_scale)
+        Kzz = rbf(Z_pos, Z_pos, k_inv_scale)
+        Kxz = rbf(X_ph, Z_pos, k_inv_scale)
+        Kzz_chol = tf.cholesky(Kzz)
+        fZ = zs.MultivariateNormalCholesky(
+            'fZ', tf.zeros([hps.N_Z], dtype=hps.dtype), Kzz_chol,
+            n_samples=n_particles)
+
+        # Model Y|Z.
+        # Mean[Y|Z] = Kxz @ inv(Kzz) @ fZ
+        # Cov[Y|Z] = Kxx - Kxz @ inv(Kzz) @ Kzx + noise_level * I
```

Comment:
Eventually we'd expect zhusuan to have a `process` module` in the distribution package. Then a GaussianProcess class  could help abstract away all these conditional queries.

Reply:
That would be nice. Although modeling and inference seem kind of coupled for Gaussian process, and I don't know whether `SparseGaussianProcess` and `RandomFeatureGaussianProcess` would be useful out of the ordinary regression context. 
I was fiddling with things like deep sparse GP before the exam week came, and I will implement the abstractions here if I get it to work / it proves to be useful.

Reply:
My imagination is something only related to modeling. The basic features of GPs, so that you can query any marginal and conditional distribution at any inputs, the returned can be directly a MultivariateNormal instance (you can choose Cholesky/full/...). That will fit nicely here. It will take some work, but we could merge this initial example first.

Reply:
@ssydasheng any thought since you are familiar with GPflow?